### DISCUSSION TITLE
Help! with slider constraints' frameA & B..

### DISCUSSION THREAD

**Discussion by EthanHermsey** [enable3d/enable3d#230] [Help / Questions]
Hello Yandeu,

I would like some help me with these rotation frames. It's breaking my brain and I don't understand it. 
Would you know how to programmatically set the axis of a slider along a vector?
I hope you can understand what I mean, and with your experience learn me how to do it properly.

[Full album](https://imgur.com/a/pbw8HjT)

I have a big cube with extendedObjects placed against it on 3 sides in the scene. The extendedObjects' y axis are pointing outwards along the normals of the cube ( drawn in red ). 
I connect each extendedObject with a slider constraint to the cube and I want them to slide outwards.
[Picture 1](https://imgur.com/2TQkxuo)

When I run the simulation, the rotation of all extendedObjects resets, so that the y-axis is pointing up.
The slider constraint's axis is ALWAYS pointing to the 'local' (1, 0, 0) direction of the cube. 
The rotation axis of the physics world (the constraints axis) seem rotated. x = z and z = x. I use a function to reverse those.
[Picture 2](https://imgur.com/k61zl9P)

I can get frameA's rotation to rotate the axis from the (1, 0, 0) axis to the desired normal vector with this code:
 ```
 const switchXZ = rotation =>{

	const x = rotation.x;
	rotation.x = rotation.z;
	rotation.z = x;

};

const axisA = this.userData.axisA.clone();                //this is the face normal of the cube

//frameA
const q = new THREE.Quaternion()
	.setFromUnitVectors(
		new THREE.Vector3( 1, 0, 0 ),                  //the direction the slider axis
		axisA                                                       //the face normal
	);

const frameA = new THREE.Euler().setFromQuaternion( q );
switchXZ( frameA );

```

But then frameB/bodyB ALSO rotates with it? That's the weird part... 
[Picture 3](https://imgur.com/XL7yglY)
I can reset that rotation again with `const frameB = frameA` so that it the extendedObjects point upward again as in the 2nd picture but I can't for the life of my figure out how to rotate frameB to point in the normal vector. I've tried the same thing as with frameA:

```
const q2 = new THREE.Quaternion()
	.setFromUnitVectors(
		new THREE.Vector3( 0, 1, 0 ),              //the extendedObjects are pointing upward
		axisA
	);

frameB = new THREE.Euler().setFromQuaternion( q2 );
switchXZ( frameB );

//Adding frameA back in?
frameB.x += frameA.x;
frameB.y += frameA.y;
frameB.z += frameA.z;
```

But this does not work. Do you know an easier solution?

Edit: I thought maybe it's the limitation of Euler, so I moved frameA back into a quaternion and multiplied it. But that also did not work.
```
let qExtra = new THREE.Quaternion().setFromEuler( frameA ); // swapped x/z, doesn't matter

//frameB
let q2 = new THREE.Quaternion()
	.setFromUnitVectors(
		new THREE.Vector3( 0, 1, 0 ),
		axisA
	)
	.multiply( qExtra );
	
frameB = new THREE.Euler().setFromQuaternion( q2 );
switchXZ( frameB );
```

**Comment by EthanHermsey** [comment]
.. I had been struggling with this for over a week and didn't get any further. One day after I ask for help, I find a solution... Using a dof constraint. When setting the angular limits to zero, the (0,1,0) axis point along the extendedObject.. 

Edit: It looks like the constraints are very weak, they quickly 'break' out of alignment, do you know if that's fixable? constraint.setBreakingImpulseThreshold() only makes it weaker, not stronger.

**Reply by yandeu** [reply]
Normally, you should not have to modify thing manually with quaternions or euler.

In my experience, you have to modify the mass if constraints are unstable. See the slider example. One cylinder has a huge mass, the other not:. 
https://github.com/enable3d/enable3d-website/blob/52693b52379427455bc2cb9212cbcb4ad88337c1/src/examples/types-of-constraints.html#L90

I will try to make an example using slider this evening or tomorrow.

**Reply by EthanHermsey** [reply]
That's what I thought.. The sourcecode for the slider looks like it would handle that rotation offset. 
I thought it was due to my case, which ís a bit weird, but when I only put 2 meshes (above eachother) in the scene and add.existing them, it does the same thing with the (1,0,0) axis.. Maybe I'm doing something wrong with the setup? Are the x,y and z axis supposed to be the same in THREE world and the physics world? Is there a setting?

The mass offset makes sense. The big cube has a very big mass, the other parts do not. When the whole thing lands on one of the other parts the constraint breaks out of alignment. Somewhere I was expecting those constraints to be completely rigid..

**Reply by yandeu** [reply]
See https://enable3d.io/examples/slider-constraints.html

**Reply by EthanHermsey** [reply]
Nice! This helped, but it's not really programmatically, it's still hardcoded. I was thinking more like clicking on a mesh and then adding a new slider on that point. On a sphere for example it would be a lot of work hardcoding rotations for each face.

I did get it to work. I changed the example and added slider constraints based on position, axis and slider length.
I personally don't like that you can see the discs rotate after adding them ( after adding the constraint, the discs rotation is reset and it then rotates toward the frameB rotation ). Is there a way to change that?



```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=1.0, user-scalable=no, viewport-fit=cover" />
    <meta http-equiv="X-UA-Compatible" content="ie=edge" />
    <title>All Constraints Types</title>
    <script src="/lib/enable3d/enable3d.framework.0.24.0.min.js"></script>
    <style>
      html, body{
        margin:0;
        width: 100vw;
        height: 100vh;
      }
      canvas{
        width: 100%;
        height: 100%;
      }
      #info-text{
        position: fixed;
      }
    </style>
  </head>

  <body>
    <div id="info-text">Slider Constraints</div>
    <script>
      const { Project, Scene3D, PhysicsLoader, THREE } = ENABLE3D

      class MainScene extends Scene3D {

        async create() {

          await this.warpSpeed('-ground')

          this.physics.debug?.enable()
          this.physics.debug?.mode(2048 + 4096)

          this.cylinderShape = { radiusTop: 0.4, radiusBottom: 0.4, height: 0.25, radiusSegments: 16 }

          this.ball = this.physics.add.sphere(
            { radius: 4, collisionFlags: 2, mass: 25 },
            { lambert: { color: 'palegreen' } }
          )

          //setup raycasting
          this.raycaster = new THREE.Raycaster();
          this.mouse = new THREE.Vector2();
          window.addEventListener('mousemove', ( event ) => {

            this.mouse.x = ( event.clientX / window.innerWidth ) * 2 - 1;
            this.mouse.y = - ( event.clientY / window.innerHeight ) * 2 + 1;

          })
          window.addEventListener('mousedown', ( event ) => {

            this.raycaster.setFromCamera( this.mouse, this.camera );
            const intersect = this.raycaster.intersectObject( this.ball )[ 0 ];
            
            if ( intersect ) {

              const arm = this.physics.add.cylinder(
                { ...this.cylinderShape, ...intersect.point }, //cylinder shape + position
                { lambert: { color: 'orange' } }
              )
              
              this.addSlider( 
                this.ball.body,
                arm.body,
                intersect.face.normal, 
                2,
                {
                  frameB: { x: Math.PI * 0.5 } //rotation so flatside of cylinder points up?
                }
              );

            }

          } )

        }

        addSlider( bodyA, bodyB, axis, length, options ){

          //frameA
          const frame = new THREE.Euler().setFromQuaternion( 
            new THREE.Quaternion()
              .setFromUnitVectors(
                new THREE.Vector3( 1, 0, 0 ),
                axis
              )
          );
          const x = frame.x;
          frame.x = frame.z;
          frame.z = x;

          this.physics.add.constraints.slider(
            bodyA, 
            bodyB, 
            {
              frameA: options.frameA || frame,
              frameB: options.frameB || frame,
              linearLowerLimit: 0,
              linearUpperLimit: length
            }
          );

        }

        update() {

          this.ball.rotation.x += 0.002
          this.ball.rotation.y += 0.006
          this.ball.body.needUpdate = true

        }

      }

      PhysicsLoader( '/lib/ammo', () => new Project( { scenes: [ MainScene ] } ) )
    </script>
  </body>
</html>
```

**Reply by EthanHermsey** [reply]
I was thinking about doing this even:


```
          this.physics.add.constraints.slider(
            this.ball.body,
            arm.body,
            {
              axis: intersect.face.normal, 
              frameB: { x: Math.PI * 0.5 }, //rotation so flatside of cylinder points up?
              linearUpperLimit: 2
            }
          );
```

```
          this.physics.add.constraints.slider = function(
            bodyA,
            bodyB,
            config = {},
            disableCollisionsBetweenLinkedBodies = true
          ){
           
            const transform = this.getTransform(bodyA.ammo, bodyB.ammo)

            const {
              frameA = {},
              frameB = {},
              axis = {},              
              linearLowerLimit: lll = 0,
              linearUpperLimit: lul = 0,
              angularLowerLimit: all = 0,
              angularUpperLimit: aul = 0
            } = config;

            if ( axis.x && axis.y && axis.z ){

              frameA = new THREE.Euler().setFromQuaternion(
              new THREE.Quaternion()
                .setFromUnitVectors(
                  new THREE.Vector3( 1, 0, 0 ),
                  axis
                )

              );
              const x = frameA.x;
              frameA.x = frameA.z;
              frameA.z = x;

            }

            const rotationA = transform.transformA.getRotation()
            rotationA.setEulerZYX(frameA.x || 0, frameA.y || 0, frameA.z || 0)
            transform.transformA.setRotation(rotationA)

            const rotationB = transform.transformB.getRotation()
            rotationB.setEulerZYX(frameB.x || 0, frameB.y || 0, frameB.z || 0)
            transform.transformB.setRotation(rotationB)

            const constraint = new Ammo.btSliderConstraint(
              bodyA.ammo,
              bodyB.ammo,
              transform.transformA,
              transform.transformB,
              true
            )

            constraint.setLowerLinLimit(lll)
            constraint.setUpperLinLimit(lul)
            constraint.setLowerAngLimit(all)
            constraint.setUpperAngLimit(aul)

            this.physicsWorld.addConstraint(constraint, disableCollisionsBetweenLinkedBodies)
            return constraint

          }
```

**Reply by EthanHermsey** [reply]
The code on the edited slider constraint is not tested and something I just typed up.
